### PR TITLE
8278165: Clarify that ZipInputStream does not access the CEN fields for a ZipEntry

### PR DESCRIPTION
**Pull request by LanceAndersen**
Hi,

Please review this update to the ZipInputStream class description to clarify that ZipInputStream walks sequentially through each Zip Entry contained within the Zip File.  As a result, the CEN header for the Zip file entries are not read resulting in ZipInputStream not having access to information that is stored within the CEN Header fields such the as the external file attributes which can be used to store optional data such as Posix Permissions.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8278165](https://bugs.openjdk.org/browse/JDK-8278165): Clarify that ZipInputStream does not access the CEN fields for a ZipEntry
 * [JDK-8293756](https://bugs.openjdk.org/browse/JDK-8293756): Clarify that ZipInputStream does not access the CEN fields for a ZipEntry (**CSR**)

### Reviewers
 * [Brian Burkhalter](https://openjdk.org/census#bpb) (@bplb - **Reviewer**) ⚠️ Review applies to [fe462d7f](https://git.openjdk.org/jdk/pull/10102/files/fe462d7f5f328843443e95afb23324fc1d04d9ab)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10102/head:pull/10102` \
`$ git checkout pull/10102`

Update a local copy of the PR: \
`$ git checkout pull/10102` \
`$ git pull https://git.openjdk.org/jdk pull/10102/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10102`

View PR using the GUI difftool: \
`$ git pr show -t 10102`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10102.diff">https://git.openjdk.org/jdk/pull/10102.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/java/util/zip/ZipInputStream.java [hunk 1]
```diff
@@ -38,9 +38,36 @@
...
- * entries.
+ * An input stream for reading compressed and uncompressed
+ * {@linkplain ZipEntry ZIP file entries} from a stream of bytes in the ZIP file
+ * format.
+ *
+ * <H2>Reading Zip File Entries</H2>
+ *
+ * The {@link #getNextEntry()} method is used to read the next ZIP file entry
+ * (Local file (LOC) header record in the ZIP format) and position the stream at
+ * the entry's file data. The file data may read using one of
```
Comment (bplb) [inline src/java.base/share/classes/java/util/zip/ZipInputStream.java:49]: of -> of the?

File: src/java.base/share/classes/java/util/zip/ZipInputStream.java [hunk 2]
```diff
@@ -38,9 +38,36 @@
...
+ * {@code ZipInputStream} read methods such
+ * as {@link #read(byte[], int, int) read} or {@link #readAllBytes() readAllBytes()}.
+ * For example:
+ * <pre>
+ *     {@code
+ *      try (FileInputStream fis = new FileInputStream(jar.toFile());
+ *              ZipInputStream zis = new ZipInputStream(fis)) {
```
Comment (bplb) [inline src/java.base/share/classes/java/util/zip/ZipInputStream.java:56]: The indentation seems weird in this code sample.

File: src/java.base/share/classes/java/util/zip/ZipInputStream.java (hunk 1 above)
Comment (LanceAndersen) [reply src/java.base/share/classes/java/util/zip/ZipInputStream.java]: fixed

File: src/java.base/share/classes/java/util/zip/ZipInputStream.java (hunk 2 above)
Comment (LanceAndersen) [reply src/java.base/share/classes/java/util/zip/ZipInputStream.java]: Reformatted

**Review by bplb — APPROVED** [review]
Looks fine.

File: src/java.base/share/classes/java/util/zip/ZipInputStream.java [hunk 3]
```diff
@@ -38,9 +38,36 @@
...
+ *           ZipInputStream zis = new ZipInputStream(fis)) {
+ *          ZipEntry ze;
+ *          while((ze= zis.getNextEntry()) != null) {
+ *             var bytes = zis.readAllBytes();
+ *             System.out.printf("Entry: %s, bytes read: %s%n", ze.getName(),
+ *                     bytes.length);
+ *          }
+ *      }
+ *    }
+ * </pre>
```
[the remark this answers is not in the document]
Comment (LanceAndersen) [reply src/java.base/share/classes/java/util/zip/ZipInputStream.java]: Converted the example to a snippet.  Thanks for the suggestion

File: src/java.base/share/classes/java/util/zip/ZipInputStream.java [hunk 4]
```diff
@@ -38,9 +38,34 @@
...
+ *             System.out.printf("Entry: %s, bytes read: %s%n", ze.getName(),
+ *                     bytes.length);
+ *          }
+ *      }
+ *    }
+ * The LOC header contains metadata about the Zip file entry. {@code ZipInputStream}
+ * does not read the Central directory (CEN) header for the entry and therefore
+ * will not have access to its metadata such as the external file attributes.
+ * {@linkplain ZipFile} may be used when the information stored within
+ * the CEN header is required.
```
Comment (AlanBateman) [inline src/java.base/share/classes/java/util/zip/ZipInputStream.java:68]: I wonder if this paragraph should be an apiNote. It's not wrong as is but it feels like a note to remind developers that this is an "input stream" for iterating through the ZIP entries. The external files attributes are in the CEN at the end of the ZIP file so this is why they aren't available.

Comment (LanceAndersen) [reply src/java.base/share/classes/java/util/zip/ZipInputStream.java:70]: > I wonder if this paragraph should be an apiNote. It's not wrong as is but it feels like a note to remind developers that this is an "input stream" for iterating through the ZIP entries. The external files attributes are in the CEN at the end of the ZIP file so this is why they aren't available.

converted to an apiNote

**Review by AlanBateman — APPROVED** [review]
I think version e15660e3 looks okay.

**Review by AlanBateman — APPROVED** [review]

**Pull request closed without merging**